packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -182,6 +182,42 @@ typedef DecoderCallback = Future<ui.Codec> Function(Uint8List bytes, {int cacheW
 ///
 /// The following image formats are supported: {@macro flutter.dart:ui.imageFormats}
 ///
+/// ## Image resolving
+///
+/// The [ImageProvider] goes through the following lifecycle to resolve an
+/// image, once the [resolve] method is called:
+///
+///   1. Create an [ImageStream] using [createStream] to return to the caller.
+///      This stream will be used to communicate back to the caller when the
+///      image is decoded and ready to display, or when an error occurs.
+///   2. Obtain the key for the image using [obtainKey].
+///      Since this method is asynchronous, but also can return synchronously,
```

Comment:
This sentence is strange. I assume you mean it's an asychnronous method, but it may throw exceptions synchronously? There is no way it can actually return anything useful synchronous?

Reply:
We use `SynchronousFuture`s for these, which look async but are actually return synchronously.

Reply:
Reworded a bit.